--- /usr/share/nvim/runtime/lua/vim/lsp.lua	2021-06-30 19:47:06.441987903 -0300
+++ lsp.lua	2021-06-11 09:55:38.700486994 -0300
@@ -228,6 +228,7 @@
     on_init         = { config.on_init, "f", true };
     settings        = { config.settings, "t", true };
     before_init     = { config.before_init, "f", true };
+    no_wait         = { config.no_wait, "b", true };
     offset_encoding = { config.offset_encoding, "s", true };
     flags           = { config.flags, "t", true };
     get_language_id = { config.get_language_id, "f", true };
@@ -597,6 +598,9 @@
 --- or "utf-32" which is the encoding that the LSP server expects. Client does
 --- not verify this is correct.
 ---
+--@param no_wait (boolean) Do not wait the language server to exit
+--- before closing NeoVim.
+---
 --@param on_error Callback with parameters (code, ...), invoked
 --- when the client operation throws an error. `code` is a number describing
 --- the error. Other arguments may be passed depending on the error kind.  See
@@ -1218,9 +1222,19 @@
   if tbl_isempty(active_clients) then
     return
   end
+
+  local pid_list = {}
+  for client_id, client in pairs(active_clients) do
+    if client.config.no_wait then
+      pid_list[#pid_list + 1] = client.rpc.handle:get_pid() 
+      active_clients[client_id] = nil
+    end
+  end
+
   for _, client in pairs(active_clients) do
     client.stop()
   end
+  require'safelsp'._call_exit(pid_list)
 
   if not vim.wait(500, function() return tbl_isempty(active_clients) end, 50) then
     for _, client in pairs(active_clients) do
@@ -1534,5 +1548,6 @@
 -- Define the LspDiagnostics signs if they're not defined already.
 require('vim.lsp.diagnostic')._define_default_signs_and_highlights()
 
+lsp._thiago_ = true
 return lsp
 -- vim:sw=2 ts=2 et
